### PR TITLE
Fix the issue of going out of bounds in the isccfg parser.

### PR DESCRIPTION
**Pull request by SandakovMM**
This problem can occur when attempting to parse an empty file.

### CODE REVIEW

**Comment by pirat89** [comment]
@SandakovMM thanks for the  contribitioun. LGTM.  @pemensik are you ok with the change?

**Review by fernflower — CHANGES_REQUESTED** [review]
Thanks for the fix! Could you please add a unit test?

**Comment by SandakovMM** [comment]
> Could you please add a unit test?
Sure

**Comment by SandakovMM** [comment]
The test currently checks if the code will not crash. I'm not sure what else we could add to the check at this point. Please share your ideas if you have any.

File: repos/system_upgrade/el7toel8/libraries/isccfg.py [hunk 1]
```diff
@@ -686,9 +686,12 @@ def find_next_key(self, cfg, index=0, end_index=-1, end_report=False):
 
         while index != -1:
             keystart = index
-            while istr[index] in self.CHAR_KEYWORD and index < end_index:
+            while index < end_index and istr[index] in self.CHAR_KEYWORD:
                 index += 1
 
+            if index >= end_index:
```
Comment (pemensik) [inline repos/system_upgrade/el7toel8/libraries/isccfg.py:692]: These changes seem correct, fixes the problem for me

File: repos/system_upgrade/el7toel8/libraries/tests/test_isccfg.py [hunk 2]
```diff
@@ -116,6 +116,8 @@
 };
 """)
 
+views_empty = isccfg.MockConfig("")
+
```
Comment (pemensik) [inline repos/system_upgrade/el7toel8/libraries/tests/test_isccfg.py:120]: I would suggest also ``MockConfig('options { include "/dev/null"; };')`` testing, so we test also empty include file. I would rename it to ``config_empty`` variable, it contains no view.

File: repos/system_upgrade/el7toel8/libraries/tests/test_isccfg.py [hunk 3]
```diff
@@ -343,5 +345,22 @@ def test_walk():
     assert 'dnssec-validation' not in state
 
 
+def test_empty_config():
+    """ Test empty configuration """
+
+    callbacks = {
```
Comment (pemensik) [inline repos/system_upgrade/el7toel8/libraries/tests/test_isccfg.py:351]: for empty file no callbacks make sense or are needed. Use just ``callbacks = {}``.

File: repos/system_upgrade/el7toel8/libraries/tests/test_isccfg.py [hunk 4]
```diff
@@ -343,5 +345,22 @@ def test_walk():
...
+        'options': cb_state,
+        'dnssec-lookaside': cb_state,
+        'dnssec-validation': cb_state,
+    }
+    state = {}
```
Comment (pemensik) [inline repos/system_upgrade/el7toel8/libraries/tests/test_isccfg.py:356]: Remove state usage, make it default to None. Unlike views test it is not needed here.

File: repos/system_upgrade/el7toel8/libraries/tests/test_isccfg.py [hunk 5]
```diff
@@ -343,5 +345,22 @@ def test_walk():
...
+
+    parser = isccfg.IscConfigParser(views_empty)
+    assert len(parser.FILES_TO_CHECK) == 1
+    cfg = parser.FILES_TO_CHECK[0]
+    parser.walk(cfg.root_section(), callbacks, state)
```
Comment (pemensik) [inline repos/system_upgrade/el7toel8/libraries/tests/test_isccfg.py:361]: Omit 3rd parameter, not needed.

**Review by pemensik — CHANGES_REQUESTED** [review]
Core change is good, I would simplify the new test function a bit. And add additional include "/dev/null"; to test also included file check.

File: repos/system_upgrade/el7toel8/libraries/tests/test_isccfg.py (hunk 4 above)
Comment (SandakovMM) [reply repos/system_upgrade/el7toel8/libraries/tests/test_isccfg.py:356]: We can probably remove the `state` variable since it can be omitted from the `walk` call

**Comment by fernflower** [comment]
@SandakovMM I could be mistaken, but I see that some comments from @pemensik have not been addressed (like https://github.com/oamg/leapp-repository/pull/1124/files/0f5b3da07b0ba3af50d88703e8620f54cd9af9be#diff-3755f648a6265f55301fda90f4333a3289285caa586c8bd8321c90d7746ae32bR361).

**Comment by fernflower** [comment]
My bad, all addressed, I will schedule tests and merge it when they are done. Thanks!

**Comment by fernflower** [comment]
Just one thing - could you please rebase?

**Comment by pirat89** [comment]
> Certainly, would you prefer to consolidate all the modifications into a single commit?

@SandakovMM  yes please :)  thanks

**Comment by fernflower** [comment]
Could you please rebase? I'd like to run regression tests for this, but I need https://github.com/oamg/leapp-repository/commit/93c6fd4f150229a01ba43ce74214043cffaf7dce (and ~10 later commits that were delivered by https://github.com/oamg/leapp-repository/pull/1110)

**Comment by SandakovMM** [comment]
Sure

**Comment by fernflower** [comment]
Either that is my bad and Friday speaking, or your branch https://github.com/SandakovMM/leapp-repository/tree/master is still missing the commits I need to enable testing :(
You need to rebase your branch against leapp-repository's master.

**Comment by SandakovMM** [comment]
My bad, seem like I've rebased agains https://github.com/AlmaLinux/leapp-repository, not https://github.com/oamg/leapp-repository. Should be fine now.

**Review by fernflower — APPROVED** [review]
lgtm, when the tests pass I'll merge it.

**Comment by fernflower** [comment]
@pirat89 but I already did that 2 days ago..

**Comment by pirat89** [comment]
@fernflower aah. so you just forgot to set the label :)
